### PR TITLE
Add network name to labels for workload endpoints

### PR DESCRIPTION
**Pull request by trev97**
## Description

This is a new feature that adds the neutron network name to the Workload endpoint's labels which allows for multi-openstack deployment policy to be created more easily than having to reference all of the individual network UUIDs across many clusters. For instance you might have a development or production network across many deployments, and being able to reference that for policy would be helpful. I've tested this in one of our clusters and it added the network_name label to the etcd datastore to all new and old Workload Endpoints. 

## Related issues/PRs
Related to [issue 9128](https://github.com/projectcalico/calico/issues/9128)

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Add a label to Workload endpoints for the name of the network of the related port. Useful for cases where there are many different OpenStack deployments with the same network name. No impact requiring code changes.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Review by nelljerram — COMMENTED** [review]
LGTM, thanks for the contribution.

Ideally we'd also have a UT.  Unfortunately that's not completely trivial because [our DB mocking](https://github.com/projectcalico/calico/blob/master/networking-calico/networking_calico/plugins/ml2/drivers/calico/test/lib.py#L243-L324) currently covers ports, subnets and security groups but not network objects.  Would you be happy to take a look at that?

**Comment by trev97** [comment]
I can certainly take a look at it at some point, however it might take a bit for it to gain priority in our sprints.

**Comment by tj90241** [comment]
We'll prioritize development of any needed tests in our upcoming sprint -- thanks for being patient and taking the time to review this PR!

**Comment by trev97** [comment]
As far as tests go, I was thinking of the following possibilities:
- Network name label should change when Network name changes
- Change in network name label should be corrected back to the Network name 
  
Are there other things to keep in mind?

**Comment by nelljerram** [comment]
@trev97 That sounds good to me.

**Comment by trev97** [comment]
@nelljerram Looking into the UT here, will this require changes to this file as well? https://github.com/projectcalico/calico/blob/master/networking-calico/networking_calico/plugins/ml2/drivers/calico/mech_calico.py#L676? 
Looking at the updating process, seems there is no method for calico to change, update, or delete the networks. Would it then make sense, as to not inflate this PR too much, to simply see that the label gets reverted when it contradicts the network names that neutron creates? Or would that still require more code changes.

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py
Comment (trev97) [inline networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py:920]: @nelljerram I haven't gotten this to work yet, my closest attempts so far being:
1. Use the `osdb_ports` change and `simulated_time_advance`.  This properly updates the `network-name` and `network-id` but leaves behind the `uid` and `creationTimestamp`
2. Use the `osdb_ports` change, `simulated_time_advance` and `context._port` + `update_port_postcommit`. This deletes the `uid` and `creationTimestamp` as expected, and makes the proper changes to the labels but writes down the security group as well. 
  
Not sure what the solution in this case is, could use some insight here.

**Comment by trev97** [comment]
@nelljerram Tests should be passing now if you would like to run the workflow.

**Comment by nelljerram** [comment]
@trev97 I need to merge #9302 (or a better version of that) first ...

**Comment by nelljerram** [comment]
While we're waiting for that, would you like to have a go at:
1. An appropriate release note - to replace "TBD" in the PR description above.
2. Doc - I think we just need a simple addition to the page at https://docs.tigera.io/calico/latest/networking/openstack/labels, for which the source is at https://github.com/tigera/docs/blob/main/calico/networking/openstack/labels.mdx.

**Comment by tj90241** [comment]
re: the docs repository: we need to work with our Github org owner to establish a fork of the Tigera docs repo such that the CLA is accounted for (I do not believe we can sign individually and there's a bit of a process here). That may take some time - we'll start turning the wheels on that effort. Thanks again for your review and continued patience!

**Comment by nelljerram** [comment]
@tj90241 @trev97 If you can now merge origin/master into this PR, I'll trigger the CI again.

**Comment by trev97** [comment]
Merged master branch.

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/lib.py
Comment (nelljerram) [inline networking-calico/networking_calico/plugins/ml2/drivers/calico/test/lib.py:629]: I think we should now rename `port_query` to `db_query` (or similar) and add comments here to say that:
- the 'port_id' case occurs when querying for IPAllocations
- the 'fixed_port_id' case occurs when querying for FloatingIPs
- the 'id' case occurs when querying for Networks.

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py
Comment (nelljerram) [inline networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py:126]: Is this a leftover from an earlier idea?

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py
Comment (nelljerram) [inline networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py:916]: Is this a leftover from an earlier idea?  I'm not sure why we are expecting this annotation.

**Review by nelljerram — CHANGES_REQUESTED** [review]
Thanks @trev97 , a few small points.

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py
Comment (trev97) [reply networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py:916]: I'm changing the network-id on the neutron port, so I expect fields related to it on the endpoint (network-id, network-name) to change.

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/lib.py
Comment (trev97) [reply networking-calico/networking_calico/plugins/ml2/drivers/calico/test/lib.py:632]: Changed name to more accurately reflect that method now.

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py
Comment (nelljerram) [reply networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py:916]: Sorry, my mistake, I had forgotten about how this annotation works, but have reacquainted myself with that now.

**Review by nelljerram — APPROVED** [review]
Thanks @trev97 !

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py
Comment (nelljerram) [reply networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py:126]: Sorry, my mistake.

However, optionally - if you don't mind another cycle of review - it would be better to update the comment above to say:
```
            # If this is an update, check that the metadata, other than labels
            # and annotations, is unchanged.
```

**Review by nelljerram — APPROVED** [review]
One last thought - up to you if you'd like to apply this and don't mind waiting for another cycle of review.

File: networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py
Comment (trev97) [reply networking-calico/networking_calico/plugins/ml2/drivers/calico/test/test_plugin_etcd.py:126]: Sure, updated comment.

**Review by nelljerram — APPROVED** [review]

**Comment by nelljerram** [comment]
@trev97 I'm happy to merge this after the next CI pass.  Are you also happy with that?

**Comment by trev97** [comment]
Sounds good.

**Comment by nelljerram** [comment]
@trev97 @tj90241 As the docs process is a little tricky for you, shall I just do it and then ask you to review?  (I think it's literally just one or two lines needed.)

**Comment by trev97** [comment]
That would be much appriciated!

**Comment by nelljerram** [comment]
> That would be much appreciated!

@trev97 @tj90241 Please could you review https://github.com/tigera/docs/pull/1702 ?

**Comment by nelljerram** [comment]
Thanks so much the contribution @trev97 !

**Pull request merged**
